### PR TITLE
feat(gen ai): add samples for aiplatform batch predict api with gemini

### PR DESCRIPTION
**Pull request by Valeriy-Burlaka**
## Description

Adds new samples for the Vertex AI "Batch Prediction" [page](https://cloud.google.com/vertex-ai/generative-ai/docs/multimodal/batch-prediction-gemini) sections:

* "Batch Prediction for Cloud Storage" [section](https://cloud.google.com/vertex-ai/generative-ai/docs/multimodal/batch-prediction-gemini#cloud-storage)
* "Batch Prediction for BigQuery" [section](https://cloud.google.com/vertex-ai/generative-ai/docs/multimodal/batch-prediction-gemini#bigquery)

## Checklist
- [x] I have followed [Contributing Guidelines from CONTRIBUTING.MD](https://github.com/GoogleCloudPlatform/golang-samples/blob/main/CONTRIBUTING.md)
- [x] **Tests** pass
- [x] **Code formatted**
- [x] **Vetting** pass
- [x] Please **merge** this PR for me once it is approved

### CODE REVIEW

File: vertexai/batch-predict/batch_predict_bq.go [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package batchpredict
+
+// [START generativeaionvertexai_batch_predict_gemini_createjob_bigquery]
```
Comment (Valeriy-Burlaka) [inline vertexai/batch-predict/batch_predict_bq.go:17]: Matches the [region name](https://github.com/GoogleCloudPlatform/python-docs-samples/pull/12731/files#diff-6264f60f05d359f717c065cd0ce26840330958a445c11992e3dc1c5a771f4a2bR24) in the Python sample.

File: vertexai/batch-predict/batch_predict_bq.go
Comment (msampathkumar) [inline vertexai/batch-predict/batch_predict_bq.go:1]: please rename the files to use prefix `gemini`. Ex: https://github.com/GoogleCloudPlatform/python-docs-samples/tree/main/generative_ai/batch_predict

File: vertexai/batch-predict/batch_predict_gcs.go [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+
+	"google.golang.org/api/option"
+	"google.golang.org/protobuf/types/known/structpb"
+)
+
+// batchPredictGCS submits a batch prediction job using GCS data source as its input
+func batchPredictGCS(w io.Writer, projectID, location string, inputURIs []string, outputURI string) error {
+	// location := "us-central1"
+	// inputURIs := []string{"gs://cloud-samples-data/batch/prompt_for_batch_gemini_predict.jsonl"}
+	// outputURI := "gs://acme-org-gen-ai/gemini-outputs/batch-predict/go-test-job-001"
```
Comment (msampathkumar) [inline vertexai/batch-predict/batch_predict_gcs.go:35]: We don't have to use unknown orgs like `acme-org-gen-ai`; You can go for syntax format like `gs://<cloud-project-name>/<cloud-bucket-name>/<file-path>/input_file.jsonl`

File: vertexai/batch-predict/batch_predict_bq.go [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+	}
+
+	req := &aiplatformpb.CreateBatchPredictionJobRequest{
+		Parent: fmt.Sprintf("projects/%s/locations/%s", projectID, location),
+		BatchPredictionJob: &aiplatformpb.BatchPredictionJob{
+			DisplayName:     jobName,
+			Model:           fmt.Sprintf("publishers/google/models/%s", modelName),
+			ModelParameters: modelParameters,
+			// Check the API reference for `BatchPredictionJob` for supported input and output formats:
+			// https://cloud.google.com/vertex-ai/docs/reference/rpc/google.cloud.aiplatform.v1#google.cloud.aiplatform.v1.BatchPredictionJob
```
Comment (msampathkumar) [inline vertexai/batch-predict/batch_predict_bq.go:62]: Nice.

**Review by telpirion — CHANGES_REQUESTED** [review]

File: vertexai/batch-predict/batch_predict_gcs.go [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+	job, err := client.CreateBatchPredictionJob(ctx, req)
+	if err != nil {
+		return err
+	}
+	fullJobId := job.GetName()
+	fmt.Fprintf(w, "submitted batch predict job %q for model %q\n", fullJobId, job.GetModel())
+	fmt.Fprintf(w, "job state is %s\n", job.GetState())
+
+	for {
+		time.Sleep(5 * time.Second)
```
[the remark this answers is not in the document]
Comment (Valeriy-Burlaka) [reply vertexai/batch-predict/batch_predict_gcs.go:91]: The batch predict job that uses our test dataset usually completes in in under ~2.5-3 minutes. The polling logic here is similar to what [we have](https://github.com/GoogleCloudPlatform/python-docs-samples/blob/main/generative_ai/batch_predict/gemini_batch_predict_gcs.py#L54) in the Python sample.

Do you mean that we overall shouldn't teach developers to do a long polling to get the job status, and use some different, more effective notification mechanism(s) instead? 
Would it help here if I added an exponential backoff, or it's better to just leave it up to a developer to figure out how they want to receive an update?

Thanks!

File: vertexai/batch-predict/batch_predict_bq.go [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+
+	"google.golang.org/api/option"
+	"google.golang.org/protobuf/types/known/structpb"
+)
+
+// batchPredictBQ submits a batch prediction job using BigQuery data source as its input
+func batchPredictBQ(w io.Writer, projectID, location string, inputURI string, outputURI string) error {
+	// location  := "us-central1"
+	// inputURI  := "bq://storage-samples.generative_ai.batch_requests_for_multimodal_input"
+	// outputURI := "bq://acme-org-dev.gemini_outputs.batch_predict_results"
```
[the remark this answers is not in the document]
Comment (Valeriy-Burlaka) [reply vertexai/batch-predict/batch_predict_bq.go]: Ah, sorry! The "ACME" name is quite popular in Google Cloud documentation — I was misled by that. Fixed now!

File: vertexai/batch-predict/batch_predict_gcs.go (hunk 2 above)
Comment (Valeriy-Burlaka) [reply vertexai/batch-predict/batch_predict_gcs.go]: Thanks, addressed!

File: vertexai/batch-predict/batch_predict_bq.go [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+			return fmt.Errorf("error: couldn't get updated job state: %w", err)
+		}
+
+		if job.GetEndTime() != nil {
+			fmt.Fprintf(w, "batch predict job finished with state %s\n", job.GetState())
+			return nil
+		} else {
+			fmt.Fprintf(w, "batch predict job is running... job state is %s\n", job.GetState())
+		}
+	}
```
[the remark this answers is not in the document]
Comment (Valeriy-Burlaka) [reply vertexai/batch-predict/batch_predict_bq.go]: Thank you. Addressed!

File: vertexai/batch-predict/batch_predict_bq.go [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+		job, err := client.GetBatchPredictionJob(ctx, &aiplatformpb.GetBatchPredictionJobRequest{
+			Name: fullJobId,
+		})
+		if err != nil {
+			return fmt.Errorf("error: couldn't get updated job state: %w", err)
+		}
+
+		if job.GetEndTime() != nil {
+			fmt.Fprintf(w, "batch predict job finished with state %s\n", job.GetState())
+			return nil
```
[the remark this answers is not in the document]
Comment (Valeriy-Burlaka) [reply vertexai/batch-predict/batch_predict_bq.go]: Thank you. Addressed!

File: vertexai/batch-predict/batch_predict_gcs.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (telpirion) [reply vertexai/batch-predict/batch_predict_gcs.go]: tl;dr: This is fine until we can identify a better option. Go ahead and resolve this conversation.

In a real world application, we would use exponential backoff, good call. However, adding backoff would make this sample more confusing. 

Ideally, the `BatchPredictionJob` struct would have a `Wait()` method, but it doesn't.

It looks like there are several polling libraries available in Go, but I'm not sure which one would be best here.

**Comment by Valeriy-Burlaka** [comment]
@telpirion , @msampathkumar , I believe that all issues with this PR have been resolved. Can you please take a final look? Thanks!

**Review by telpirion — APPROVED** [review]

**Pull request merged**